Fixed search for fx files when no ``mip`` is given

schlunma opened the pull request:
<!--
    Thank you for contributing to our project!

    Please do not delete this text completely, but read the text below and keep
    items that seem relevant. If in doubt, just keep everything and add your
    own text at the top, a reviewer will update the checklist for you.

-->

This PR fixes the automatic retrieval of fx files when no `mip` is specified.

I agree with @sloosvel that a predefined order for the mip tables is not useful due to the many different project tables. Therefore I just implemented an alphabetic ordering, i.e., the first `mip` that has data is automatically selected. To bypass this, the user has to define a specific `mip` in the recipe. I added a paragraph to the documentation that describes this process.

This should go into 2.3.1 since this fixes a bug that (1) lets the test fails and (2) might lead to the tool wrongly complaining about missing data. #1169 only disabled the tests and didn't actually fix the bug.

<!--
    Please describe your changes here, especially focusing on why this pull
    request makes ESMValCore better and what problem it solves.

    Before you start, please read our contribution guidelines: https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html

    Please fill in the GitHub issue that is closed by this pull request,
    e.g. Closes #1903
-->

Closes #1159.

Link to documentation: https://esmvaltool--1216.org.readthedocs.build/projects/ESMValCore/en/1216/recipe/preprocessor.html#fx-variables-as-cell-measures-or-ancillary-variables

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

<!-- The next two lines turn the 🛠 and 🧪 below into hyperlinks -->
[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

zklaus commented:
The issue with this approach is that there are experiments, where there are two or more fx files for the same variable and it is impossible to determine a generic right order, e.g. [here](http://clipc-services.ceda.ac.uk/dreq/u/e9b495e2-5989-11e6-a4be-ac72891c3257.html). I think https://github.com/ESMValGroup/ESMValCore/issues/1159#issuecomment-857770098 is relevant. In other words, just specify the correct table in the recipe.

zklaus reviewed (changes requested):
I don't think any generic ordering can meaningfully resolve existing ambiguity in the table selection.

zklaus commented:
I do like the documentation improvements. Let's work that into 2.4.0.

schlunma commented:
> I don't think any generic ordering can meaningfully resolve existing ambiguity in the table selection.

I agree, that's why I added this note to the documentation:

```rst
.. note::
   Some fx variables exist in more than one table (e.g., ``volcello`` exists in
   CMIP6's ``Ofx`` and ``Omon`` table or ``sftgif`` exists in CMIP6's ``fx``
   and ``IyrAnt`` table). If a specific table is desired, this needs to be
   specified by ``mip``, otherwise the table is selected based on the
   alphabetical order and data availability.
```

Do you want the tool to fail if there's is ambiguity? Or issue a warning?

I still thinks this should go into 2.3.1 since this merely fixes a bug that was introduced in 2.3.0.

zklaus commented:
I think the tool should fail if there is ambiguity.

As for the bug, are you referring to #1159? If so, in my view, that doesn't reflect the introduction of a new bug, but the uncovering of an existing bug in the test, namely the ambiguity. The solution seems clear: specify the correct table in the recipe used in the test. I am not sure why @sloosvel's comment there went unanswered.

If you mean another bug, I am sorry I missed it. In that case, could you point me to the right issue, please?

schlunma commented:
The tests were designed for the pre-#999 situation and did not fail there. This old implementation did only check tables that included `fx` in their name, so the new version is definitely an improvement here. However, the point is that the tests didn't fail on any machine before #999 and failed after it on some machines.

#1169 "solved" this by disabling the tests. I don't think that's the way to go. This PR fixes the tests by introducing a machine-independent ordering for the tables.

I agree with you that in case of ambiguity the tool should fail, but this would be a new feature and should go into v2.4 (I can open another PR for that).

zklaus commented:
It seems we agree that the bug is in the test. We also agree that disabling the test is not really the right way to address this. But there seems to be a clear solution: fix the test. Why would we want to change the behavior of the tool to work around a buggy test?

schlunma commented:
- Added [`break`](https://github.com/ESMValGroup/ESMValCore/pull/1216/files#diff-b675431b6cf759c61f7867680472aabc0304475fd666ed514ca0365bb87c2a47R366) statement: This was missing in #999 (see https://github.com/ESMValGroup/ESMValCore/issues/1159#issuecomment-856836940 for further information.
-Added [`sorted`](https://github.com/ESMValGroup/ESMValCore/pull/1216/files#diff-b675431b6cf759c61f7867680472aabc0304475fd666ed514ca0365bb87c2a47R387-R391): This was not necessary pre-#999 but is necessary now to ensure a consistent behavior across machines. This can be removed once we make the tool fail in case of ambiguity.
- Changed message in [`RecipeError`](https://github.com/ESMValGroup/ESMValCore/pull/1216/files#diff-b675431b6cf759c61f7867680472aabc0304475fd666ed514ca0365bb87c2a47R401-R407): This only changes the message in case a `mip` is given. I can remove this if you like, but I think there is no harm in having this.

sloosvel commented:
I think the changes make the documentation much better. 

I agree that that `break` should not have been removed, it went over my head. And in case of ambiguity (which I have yet to see an exp with the same fx var in multiple mips, but who knows), specifying the mip solves the issue so it's much simpler than trying to come up with a search order that would fit all projects.

valeriupredoi commented:
@sloosvel - @zklaus provides an example of exactly that - `areacellg` in two mips, [here](http://clipc-services.ceda.ac.uk/dreq/u/e9b495e2-5989-11e6-a4be-ac72891c3257.html) and there are others (I believe even `volcello` that could be `fx` or `Ofx` if it's time variant or not, Klaus correct me if I'm wrong pls). I like this approach but I think we need the run to stop if it encounters multiple mips for the same fx variable, and prompt the user to make up their mind and add valid mip. This saves us from any future headaches due to ESGF data standards variations. Note the current lazy approach to not produce these sort of variables for all exps and mips and shortshifting by making use of just a small data production eg fx vars are produced for one ensemble only or one experiment only

zklaus commented:
> (I believe even `volcello` that could be `fx` or `Ofx` if it's time variant or not, Klaus correct me if I'm wrong pls).

Almost right: `volcello` can be in one or more of `Ofx`, `Omon`, `Oyr`, and `Odec`.

schlunma commented:
All right, I will alter the code so that it raises an error if multiple `mips` are encountered that include a given fx variable. I still believe that this should go into the release 2.3.1, since it's clearly a fix of a broken feature.

valeriupredoi commented:
cheers @zklaus and cheers @schlunma too, and I too believe this should be part of 2.3.1 but Klaus is the release bossman :+1:

zklaus commented:
Alright, I am not against putting this in 2.3.1, but it still needs a bit of work (I spotted some things in the documentation at least and I think we don't have any formal review yet), so it might come down to a question of timing.

schlunma commented:
Alright, now the tool raises an error if an fx variable is found in multiple tables. I also adapted the documentation accordingly.

valeriupredoi commented on doc/recipe/preprocessor.rst:
```diff
@@ -180,22 +180,68 @@ To get an overview on data fixes and how to implement new ones, please go to
+If ``mip`` is not given, ESMValTool will search for the fx variable in all
+available tables of the specified project.
+
+.. note::
```
```suggestion
.. warning::
```

zklaus commented on doc/recipe/preprocessor.rst:
```diff
@@ -180,22 +180,68 @@ To get an overview on data fixes and how to implement new ones, please go to
-    - ``mask_landseaice``
-    - ``volume_statistics``
-    - ``weighting_landsea_fraction``
+    - area_statistics_ (``areacella``, ``areacello``)
+    - :ref:`mask_landsea<land/sea/ice masking>` (``sftlf``, ``sftof``)
+    - :ref:`mask_landseaice<ice masking>` (``sftgif``)
+    - volume_statistics_ (``volcello``)
+    - :ref:`weighting_landsea_fraction<land/sea fraction weighting>` (``sftlf``, ``sftof``)
```
Perhaps format as table with columns preprocessor, applicable fx variables?
Also, why the use of `:ref:` only for some of the preprocessor?

schlunma replied:
:ref: is only necessary for those that are not section titles yet. If you have a look at the compiled doc, you don't see any difference, all preprocessors are properly linked.

zklaus replied:
You may want to read https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#cross-referencing-arbitrary-locations. To quote:
> Using ref is advised over standard reStructuredText links to sections (like `Section title`_) because it works across files, when section headings are changed, will raise warnings if incorrect, and works for all builders that support cross-references.

zklaus commented on doc/recipe/preprocessor.rst:
```diff
@@ -180,22 +180,68 @@ To get an overview on data fixes and how to implement new ones, please go to
 ====================================================
-The following preprocessor may require the use of ``fx_variables`` 
-to be able to perform the computations:
+The following preprocessor may require the use of ``fx_variables`` to be able
```
s/preprocessor/preprocessors/

zklaus commented on doc/recipe/preprocessor.rst:
```diff
@@ -180,22 +180,68 @@ To get an overview on data fixes and how to implement new ones, please go to
+      volcello:
+
+or calling the variables and adding specific variable parameters (the key-value
+pair may be as specific as a CMOR variable can permit):
```
This line doesn't read quite right. Could you double check?

zklaus commented on doc/recipe/preprocessor.rst:
```diff
@@ -180,22 +180,68 @@ To get an overview on data fixes and how to implement new ones, please go to
+.. code-block:: yaml
+
+    fx_variables:
+      areacello:
+        mip: Omon
+      volcello:
+        mip: fx
```
```suggestion
      areacello:
        mip: Ofx
      volcello:
        mip: Omon
```
The area of ocean cells typically stays the same, but changes in sea surface height vary the thickness and thus the volume.

zklaus commented on doc/recipe/preprocessor.rst:
```diff
@@ -180,22 +180,68 @@ To get an overview on data fixes and how to implement new ones, please go to
+
+.. code-block:: yaml
+
+    fx_variables: [{'short_name': 'areacello', 'mip': 'Omon'}, {'short_name': 'volcello', 'mip': 'fx'}]
```
```suggestion
    fx_variables: [{'short_name': 'areacello', 'mip': 'Ofx'}, {'short_name': 'volcello', 'mip': 'Omon'}]
```

zklaus commented on doc/recipe/preprocessor.rst:
```diff
@@ -180,22 +180,68 @@ To get an overview on data fixes and how to implement new ones, please go to
+
+.. note::
+   Some fx variables exist in more than one table (e.g., ``volcello`` exists in
+   CMIP6's ``Ofx``, ``Oyr``, ``Odec`` and ``Omon`` tables or ``sftgif`` exists
```
```suggestion
   CMIP6's ``Ofx``, ``Oyr``, ``Odec``, and ``Omon`` tables or ``sftgif`` exists
```
Oxford comma

zklaus commented on doc/recipe/preprocessor.rst:
```diff
@@ -180,22 +180,68 @@ To get an overview on data fixes and how to implement new ones, please go to
+.. note::
+   Some fx variables exist in more than one table (e.g., ``volcello`` exists in
+   CMIP6's ``Ofx``, ``Oyr``, ``Odec`` and ``Omon`` tables or ``sftgif`` exists
+   in CMIP6's ``fx``, ``LImon``, ``IyrGre`` and ``IyrAnt`` tables). In these
```
```suggestion
   in CMIP6's ``fx``, ``LImon``, ``IyrGre``, and ``IyrAnt`` tables). In these
```

valeriupredoi reviewed (approved):
excellent work, cheers @schlunma :beer: One nag for me, I'd change that from a Note to a Warning :+1:

zklaus commented on doc/recipe/preprocessor.rst:
```diff
@@ -180,22 +180,68 @@ To get an overview on data fixes and how to implement new ones, please go to
+.. note::
+   Some fx variables exist in more than one table (e.g., ``volcello`` exists in
+   CMIP6's ``Ofx``, ``Oyr``, ``Odec`` and ``Omon`` tables or ``sftgif`` exists
+   in CMIP6's ``fx``, ``LImon``, ``IyrGre`` and ``IyrAnt`` tables). In these
+   cases, ``mip`` needs to be specified, otherwise an error is raised.
```
Perhaps it's better to say "In case of ambiguity, ..." because sometimes the variable may be in different tables for different datasets, but every dataset individually is unambiguous.

schlunma replied:
No - this is exactly what the code does, see also my comment below.

zklaus commented on doc/recipe/preprocessor.rst:
```diff
@@ -180,22 +180,68 @@ To get an overview on data fixes and how to implement new ones, please go to
+   in CMIP6's ``fx``, ``LImon``, ``IyrGre`` and ``IyrAnt`` tables). In these
+   cases, ``mip`` needs to be specified, otherwise an error is raised.
 
 The preprocessor step ``add_fx_variables`` loads the required ``fx_variables``,
```
I think we should either add a `:ref:` to `add_fx_variables` or avoid explicit mentioning of it (something along the lines of "...an internal preprocessor step...") to avoid confusion.

schlunma replied:
I slightly modified this by adding "internally" and "automatically". Note that I only copy-pasted these parts of the documentation, they were already present in the old version.

zklaus commented on doc/recipe/preprocessor.rst:
```diff
@@ -180,22 +180,68 @@ To get an overview on data fixes and how to implement new ones, please go to
+   cases, ``mip`` needs to be specified, otherwise an error is raised.
 
 The preprocessor step ``add_fx_variables`` loads the required ``fx_variables``,
 checks them against CMOR standards and adds them either as ``cell_measure``
```
Link to [CF](http://cfconventions.org/cf-conventions/cf-conventions.html#cell-measures) and or [Iris](https://scitools-iris.readthedocs.io/en/latest/generated/api/iris/coords.html#iris.coords.CellMeasure) cell measures?

zklaus commented on doc/recipe/preprocessor.rst:
```diff
@@ -180,22 +180,68 @@ To get an overview on data fixes and how to implement new ones, please go to
 
 The preprocessor step ``add_fx_variables`` loads the required ``fx_variables``,
 checks them against CMOR standards and adds them either as ``cell_measure``
 or ``ancillary_variable`` inside the cube data. This ensures that the
```
Link to [CF](http://cfconventions.org/cf-conventions/cf-conventions.html#ancillary-data) or [Iris](https://scitools-iris.readthedocs.io/en/latest/generated/api/iris/coords.html#iris.coords.AncillaryVariable) ancillary variables?

zklaus commented on doc/recipe/preprocessor.rst:
```diff
@@ -180,22 +180,68 @@ To get an overview on data fixes and how to implement new ones, please go to
 
```
Adding an explicit anchor (something like, but better than `.. _fx ancillary cell measures:`) will make it easier in the future to keep the links intact.

schlunma replied:
I don't get this comment.There is an anchor right there which I'm also referring to? `.. _Fx variables as cell measures or ancillary variables:`

zklaus replied:
This relates to the use of `:ref:` later on. Sorry, the terminology is a bit convoluted in this space: You are linking directly to the title of the section using what naked restructured text calls [an implicit hyperlink target](https://docutils.sourceforge.io/docs/user/rst/quickref.html#hyperlink-targets) and I think I saw referred to in a sphinx context as an automatic label.

The problem with this is, when at some point in the future we change the heading, all the links break. Also, you cannot refer to this from a different file in the same document.

If instead we put an extra anchor or label here, we can refer to it from anywhere and the validity of the links is ensured by sphinx.

schlunma replied:
But there IS an anchor:

```rst
...
To get an overview on data fixes and how to implement new ones, please go to
:ref:`fixing_data`.

.. _Fx variables as cell measures or ancillary variables:

Fx variables as cell measures or ancillary variables
====================================================
The following preprocessor may require the use of ``fx_variables`` 
to be able to perform the computations:
...
```

It may be hidden in the diff since it's already present in the current code.

zklaus replied:
You are right, sorry! I did indeed miss the anchor because it didn't show up in the diff and also because it is exactly the same text as the title. I think it would be better if it was different, but no need to change it here.

zklaus commented on doc/recipe/preprocessor.rst:
```diff
@@ -377,10 +423,13 @@ or alternatively:
             ]
 
+More details on the argument ``fx_variables`` and its default values are given
+:ref:`here<Fx variables as cell measures or ancillary variables>`.
```
See comment on explicit anchor at the top.

zklaus commented on doc/recipe/preprocessor.rst:
```diff
@@ -458,10 +502,13 @@ or alternatively:
             ]
 
+More details on the argument ``fx_variables`` and its default values are given
+:ref:`here<Fx variables as cell measures or ancillary variables>`.
```
See comment on explicit anchor at the top.

zklaus commented on doc/recipe/preprocessor.rst:
```diff
@@ -514,6 +560,9 @@ or alternatively:
           fx_variables: [{'short_name': 'sftgif', 'exp': 'piControl'}]
 
+More details on the argument ``fx_variables`` and its default values are given
+:ref:`here<Fx variables as cell measures or ancillary variables>`.
```
See comment about explicit anchor on top.

zklaus commented on doc/recipe/preprocessor.rst:
```diff
@@ -1420,40 +1469,9 @@ Note that this function is applied over the entire dataset. If only a specific
-variables and pass them to the function for loading and processing.
+The optional ``fx_variables`` argument specifies the fx variables that the user
+wishes to input to the function. More details on this are given
+:ref:`here<Fx variables as cell measures or ancillary variables>`.
```
See comment on explicit anchor on top.

zklaus commented on doc/recipe/preprocessor.rst:
```diff
@@ -1496,42 +1514,10 @@ This function takes the argument: ``operator``, which defines the operation to
+works best when the ``fx_variables`` provide the cell volume. The optional
+``fx_variables`` argument specifies the fx variables that the user wishes to
+input to the function. More details on this are given
+:ref:`here<Fx variables as cell measures or ancillary variables>`.
```
Explicit anchor comment.

zklaus commented on esmvalcore/_recipe.py:
```diff
@@ -349,26 +349,41 @@ def _add_fxvar_keys(fx_info, variable):
-                logger.debug("Found fx variables '%s':\n%s",
-                             fx_info['short_name'], pformat(fx_files))
-    return found_mip, fx_info, fx_files
+def _search_fx_mip(tables, variable, fx_info, config_user):
```
Hm. I am not sure we are on the same page here. The ambiguity does not come from the variable being defined in several tables. It comes into play when a model delivers it in several tables. For example, for CMIP6 `volcello` can be delivered in `Ofx` or (`Omon` and `Odec`). If the model delivers it only in `Ofx`, that's fine and unambiguous. If it delivers it in `Ofx` and at least one of the other two, that's an error. Delivering it for both `Omon` and `Odec` is ok; I am not sure about delivering it only for, say, `Omon`.

schlunma replied:
Yes, I think we are and I also understand your point.

To implement it like you describe it here a whole bunch of custom rules for every fx variable for every project has to be implemented. I thought we agreed that this is not really feasible. I think simply specifying the `mip` in these cases is by far the simplest solution that works for every project.

zklaus replied:
I don't think we would need a lot of custom rules. What I propose is basically to send the data finder out and do its worst and if it finds more than one candidate, then we barf.

schlunma replied:
The way you described it here would mean custom rules:

> For example, for CMIP6 `volcello` can be delivered in `Ofx` or (`Omon` and `Odec`). If the model delivers it only in `Ofx`, that's fine and unambiguous. If it delivers it in `Ofx` and at least one of the other two, that's an error. Delivering it for both `Omon` and `Odec` is ok; I am not sure about delivering it only for, say, `Omon`.

How else should the tool know that `Omon` and `Odec` is ok, but `Omon` and `Ofx` is not?

Raising an error when files are found in more than one should not be an issue, I can do that. But bear in mind that this could mean that for different datasets fx variables from different tables (e.g., `IyrAnt` and `fx`) might be mixed. Is this really what we want?

zklaus replied:
You are right, the tool would not know this. I should have phrased that better. It is true that the perfect solution, where the tool understands exactly which combinations are ok would require such rules. But my goal was to get an 80% solution by complaining only when there is actual ambiguity in the available data.

As for the second point, we absolutely want to support fx files coming from different tables for different datasets. The abovementioned `volcello` is a good example. Most models will have that in `Ofx` but [some](https://esgf-data.dkrz.de/search/cmip6-dkrz/?variable_id=volcello&table_id=Omon) have it in `Omon`.

schlunma replied:
Alright, will implement this.

zklaus reviewed (changes requested):
Sorry, only made it just to the actual code; have to continue later. Anyway, I think the most important comment is the last one.

zklaus commented on doc/recipe/preprocessor.rst:
```diff
@@ -180,22 +180,80 @@ To get an overview on data fixes and how to implement new ones, please go to
+      areacello:
+      volcello:
+
+or by explicitly adding variable parameters (the key-value pair may be as
+specific as a CMOR variable can permit):
```
I don't understand what this sentence means. Could you elaborate?

schlunma replied:
I'll try to rephrase (just copy-pasted this code from the old version).

zklaus reviewed (changes requested):
Sorry if something is very weird; I had some trouble with the github web interface.

schlunma commented:
I don't know why the tests are failing, I though these

```
TypeError: cannot compare cftime.DatetimeGregorian(300, 1, 16, 12, 0, 0, 0) and datetime.datetime(300, 1, 16, 12, 0) (different calendars)
```

errors were fixed by #1173?

valeriupredoi commented:
pls merge `main` in here, if you done that since yesterday afternoon then the test container needs to rebuild with the new environment from #1197 (I belive it's done that last night) - give it a day, it'll do it, if not bugger @bouweandela see if it's stuck. Note that you can run the tests locally too and see it works

schlunma commented:
After the merge of `main` the error got more interesting :smile: 

```
ValueError: Tag 'andela_bouwe' does not exist in section 'authors'
```

valeriupredoi commented:
yes, @bouweandela is famous for causing SegFaults :rofl: It's a SegFault, Manu, am rerunning it just for you :grin:

schlunma commented:
The error didn't disappear after 3 restarts, so I guess there is something wrong here... Tests run fine locally.

valeriupredoi commented:
hang on, lemme investigate :mag

valeriupredoi commented:
running tests locally I am getting this:
```
E       assert "Requested fx variable 'volcello' for dataset 'CanESM5' of project 'CMIP6' is available in more than one CMOR table for 'CMIP6': ['Ofx', 'Oyr', 'Odec', 'Omon']" in "Requested fx variable 'volcello' for dataset 'CanESM5' of project 'CMIP6' is available in more than one CMOR table for 'CMIP6': ['Oyr', 'Ofx', 'Odec', 'Omon']"
E        +  where "Requested fx variable 'volcello' for dataset 'CanESM5' of project 'CMIP6' is available in more than one CMOR table for 'CMIP6': ['Oyr', 'Ofx', 'Odec', 'Omon']" = <[RecursionError('maximum recursion depth exceeded while getting the repr of an object') raised in repr()] RecipeError object at 0x7fc19b08a4c0>.message

tests/integration/test_recipe.py:2799: AssertionError
```
and no SegFault after 2 runs, after updating to Python 3.9.6 too

schlunma commented:
Oh, that's again related to the machine-dependent ordering. Lemme change that real quick.

valeriupredoi commented:
you know what? That recursion depth error might have given me a massive clue why we get all those SegFaults anyway

valeriupredoi commented:
> Oh, that's again related to the machine-dependent ordering. Lemme change that real quick.

you not running Ubuntu too on your machine?

schlunma commented:
Was doing the tests on mistral which runs on Red Hat Enterprise Linux

valeriupredoi commented:
weird coz I think that Circle runs on CentOS

schlunma commented:
Thanks so much V., that seems to have done the trick! :tada: 

@zklaus Any further comments from your side? I hope I addressed them all.

valeriupredoi commented:
so funny - now the tests pass fine and no more SegFault - I am going to get to the bottom of those SegFaults so help me Lord Vader! :grin:

schlunma commented:
Haha, I hope you will!

zklaus commented:
Nice work, @schlunma ! :beers:
